packages/perspective-viewer-d3fc/src/ts/types.ts

```diff
@@ -0,0 +1,225 @@
+// ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
+// ┃ ██████ ██████ ██████       █      █      █      █      █ █▄  ▀███ █       ┃
+// ┃ ▄▄▄▄▄█ █▄▄▄▄▄ ▄▄▄▄▄█  ▀▀▀▀▀█▀▀▀▀▀ █ ▀▀▀▀▀█ ████████▌▐███ ███▄  ▀█ █ ▀▀▀▀▀ ┃
+// ┃ █▀▀▀▀▀ █▀▀▀▀▀ █▀██▀▀ ▄▄▄▄▄ █ ▄▄▄▄▄█ ▄▄▄▄▄█ ████████▌▐███ █████▄   █ ▄▄▄▄▄ ┃
+// ┃ █      ██████ █  ▀█▄       █ ██████      █      ███▌▐███ ███████▄ █       ┃
+// ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
+// ┃ Copyright (c) 2017, the Perspective Authors.                              ┃
+// ┃ ╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌ ┃
+// ┃ This file is part of the Perspective library, distributed under the terms ┃
+// ┃ of the [Apache License 2.0](https://www.apache.org/licenses/LICENSE-2.0). ┃
+// ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
+
+import { IPerspectiveViewerElement } from "@finos/perspective-viewer";
+
+export interface Element {
+    msMatchesSelector(selectors: string): boolean;
+}
+
+export interface Chart {
+    (container: any, settings: any): void;
+    plugin: {
+        name: string;
+        category: string;
+        max_cells: number;
+        max_columns: number;
+        render_warning: boolean;
+        initial: {
+            type?: string;
+            count?: number;
+            names: string[];
+        };
+        selectMode?: string;
+    };
+}
+
+export type PadUnit = "percent" | "domain";
+export type Pad = [number, number];
+
+export type GetSetReturn<T, Arg, Func> = T extends undefined
```

Comment:
d3 seems to do this with function overloads instead of a unified type, see e.g. [d3-shape.d.ts](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/d3-shape/index.d.ts). Since we're writing and immediately implementing our function, this may mean we need to rewrite the component into explicit getter/setter tuples (usually pairs, but a triple in the code below), or perhaps just add a separate .d.ts beside the original js file.
```typescript
    /**
     * Returns the current data comparator, which defaults to null.
     */
    sort(): ((a: Datum, b: Datum) => number) | null;
    /**
     * Sets the data comparator to the specified function and returns this pie generator.
     *
     * If both the data comparator and the value comparator are null, then arcs are positioned in the original input order.
     * Otherwise, the data is sorted according to the data comparator, and the resulting order is used. Setting the data comparator implicitly sets the value comparator to null.
     *
     * Sorting does not affect the order of the generated arc array which is always in the same order as the input data array; it merely affects the computed angles of each arc.
     * The first arc starts at the start angle and the last arc ends at the end angle.
     *
     * @param comparator A compare function takes two arguments a and b, each elements from the input data array.
     * If the arc for a should be before the arc for b, then the comparator must return a number less than zero;
     * if the arc for a should be after the arc for b, then the comparator must return a number greater than zero;
     * returning zero means that the relative order of a and b is unspecified.
     */
    sort(comparator: (a: Datum, b: Datum) => number): this;
    /**
     * Sets the data comparator to null and returns this pie generator.
     *
     * If both the data comparator and the value comparator are null, then arcs are positioned in the original input order.
     *
     * @param comparator null, to set the pie generator to use the original input order or use the sortValues comparator, if any.
     */
    sort(comparator: null): this;
```

Reply:
I need to play around a little bit more with it, but so far to it looks like method overloading is a much better option than using `GetSetReturn`.  (thanks @ada-x64!). 

```typescript

interface AxisFactory {
    (data: any[]): any;
    memoValue(): Axis;
    memoValue(a: Axis): this;

    excludeType(): () => AxisTypeValues;
    excludeType(nextExcludeType?: AxisTypeValues): this;
    // ... everything else
}

const AxisFactory = (settings: Settings): AxisFactory => {
      const _factory = (data): AxisFactoryContent = { ... };

      _factory.memoValue = (nextMemoValue?: Axis): Axis | AxisFactory => {
        if (!nextMemoValue) {
            return memoValue;
        }

        memoValue = nextMemoValue;
        return _factory;
    };

    _factory.excludeType = (nextExcludeType?: AxisTypeValues) => {
        if (!nextExcludeType) {
            return excludeType;
        }

        excludeType = nextExcludeType;
        return _factory;
    };
}
```